shell/platform/windows/keyboard_key_handler.h

```diff
@@ -0,0 +1,173 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_SHELL_PLATFORM_WINDOWS_KEYBOARD_KEY_HANDLER_H_
+#define FLUTTER_SHELL_PLATFORM_WINDOWS_KEYBOARD_KEY_HANDLER_H_
+
+#include <deque>
+#include <memory>
+#include <string>
+
+#include "flutter/shell/platform/common/client_wrapper/include/flutter/basic_message_channel.h"
+#include "flutter/shell/platform/common/client_wrapper/include/flutter/binary_messenger.h"
+#include "flutter/shell/platform/windows/keyboard_handler_base.h"
+#include "flutter/shell/platform/windows/public/flutter_windows.h"
+#include "rapidjson/document.h"
+
+namespace flutter {
+
+class FlutterWindowsView;
+
+// Handles key events.
+//
+// This class detects whether an incoming event is a redispatched one,
+// dispatches native events to delegates and collect their responses,
+// and redispatches events unhandled by Flutter back to the system.
+// See |KeyboardHook| for more information about dispatching.
+//
+// The exact behavior to handle events are further forwarded into
+// delegates. See |KeyboardKeyHandlerDelegate| and its subclasses.
+class KeyboardKeyHandler : public KeyboardHandlerBase {
+ public:
+  // An interface for concrete definition of how to asynchronously handle key
+  // events.
+  class KeyboardKeyHandlerDelegate {
+   public:
+    // Defines how to how to asynchronously handle key events.
+    //
+    // |KeyboardHook| should invoke |callback| with the response (whether the
+    // event is handled) later for exactly once.
+    virtual void KeyboardHook(int key,
+                              int scancode,
+                              int action,
+                              char32_t character,
+                              bool extended,
+                              bool was_down,
+                              std::function<void(bool)> callback) = 0;
+
+    virtual ~KeyboardKeyHandlerDelegate();
+  };
+
+  using EventRedispatcher =
+      std::function<UINT(UINT cInputs, LPINPUT pInputs, int cbSize)>;
+
+  // Create a KeyboardKeyHandler and specify where to redispatch events.
+  //
+  // The |redispatch_event| is typically |SendInput|.
+  explicit KeyboardKeyHandler(EventRedispatcher redispatch_event);
+
+  ~KeyboardKeyHandler();
+
+  // Add a delegate that handles events received by |KeyboardHook|.
+  void AddDelegate(std::unique_ptr<KeyboardKeyHandlerDelegate> delegate);
+
+  // Handles a key event.
+  //
+  // Returns whether this handler claims to handle the event, which is true if
+  // the event is a native event, or false if the event is a redispatched one.
```

Comment:
Do you mean "native", or "new"? (I'm not clear what you mean by "native" here.  Maybe you mean "non-synthesized"?)

Reply:
Yes. I've rephrased the comment here.